torch/csrc/jit/script/compiler.cpp

```diff
@@ -72,14 +176,110 @@ struct to_ir {
+    auto save_true = emitSingleIfBranch(
+        true_block, stmt.trueBranch(), &mutated_parent_values);
+    auto save_false = emitSingleIfBranch(
+        false_block, stmt.falseBranch(), &mutated_parent_values);
+
+    std::vector<std::string> sorted_mutations(
+        mutated_parent_values.begin(), mutated_parent_values.end());
+    std::sort(sorted_mutations.begin(), sorted_mutations.end());
+
+    // Register outputs in each block
+    for (const auto& x : sorted_mutations) {
+      true_block->registerOutput(save_true->getVar(x, stmt));
+    }
+    for (const auto& x : sorted_mutations) {
+      false_block->registerOutput(save_false->getVar(x, stmt));
+    }
+
+    // Add op outputs
+    for (const auto& x : sorted_mutations) {
+      environment_stack->setVar(x, n->addOutput());
+    }
   }
 
   void emitWhile(const While& stmt) {
-    // TODO: add support for control flow ops
-    throw ErrorReport(stmt) << "Control flow is not supported yet.";
+    // Emits a loop operators conforming to the semantics specified at
+    // https://github.com/onnx/onnx/blob/master/docs/Operators.md#experimental-loop
+    // TODO: implement scan_outputs
+
+    Value* trip_count_dummy = emitConst(0, "i")[0];
+    Value* cond_value = emitExpr(stmt.cond(), 1)[0];
+
+    Node* n = def.graph->insertNode(def.graph->create(kLoop, 0));
+    n->addInput(trip_count_dummy);
+    n->addInput(cond_value);
+    auto* body_block = n->addBlock();
+    // Trip count required by spec. Since this is a while loop, we do not
+    // provide access to this from user code
+    // TODO: it seems like we should implement a `for` loop as well, otherwise
+    // we'll probably have to pattern match iteration number machinery in user
+    // code to conform to the spec
+    body_block->addInput(); // Iteration num
+    body_block->addInput(); // Condition
+    size_t skip_inputs_num = 2;
+
+    {
+      environment_stack =
+          std::make_shared<Environment>(body_block, environment_stack);
+      WithInsertPoint guard(*def.graph, body_block);
+      emitStatements(stmt.body());
+
+      // Also emit the conditional
+      Value *body_cond_value = emitExpr(stmt.cond(), 1)[0];
+      body_block->registerOutput(body_cond_value);
+
+      // Remove inputs for values that did not mutate within the
+      // block
+      environment_stack->deleteExtraInputs(skip_inputs_num);
+
+      // Add block outputs
+      auto& curr_frame = *environment_stack;
+      for (const auto& x : curr_frame.captured_inputs) {
+        body_block->registerOutput(curr_frame.value_table[x]);
+      }
+
+      auto preserve_captured_inputs = curr_frame.captured_inputs;
+
+      // Drop out of block environment
+      environment_stack = environment_stack->next;
+
+      // Add op inputs
+      for (const auto& x : preserve_captured_inputs) {
+        n->addInput(environment_stack->getVar(x, stmt));
+        environment_stack->setVar(x, n->addOutput());
+      }
+    }
```

Comment:
I think a nicer order (that would also let you avoid a copy of `captured_inputs`) would be this:
```cpp
auto next_frame = environment_stack->next;
for (const auto& x : curr_frame.captured_inputs) {
  n->addInput(next_frame->getVar(x, stmt));
  next->setVar(x, n->addOutput());
}
environment_stack = next_frame;
```